shared/domains/chelonia/internals.js

```diff
@@ -106,18 +131,51 @@ sbp('sbp/selectors/register', {
         state._vm.props[v.key] = v.value
       },
       [GIMessage.OP_KEY_ADD] (v: GIOpKeyAdd) {
-        // TODO: implement this. consider creating a function so that
-        //       we're not duplicating code in [GIMessage.OP_CONTRACT]
-        // if (!state._vm.authorizedKeys) state._vm.authorizedKeys = []
-        // state._vm.authorizedKeys.push(v)
+        const keys = { ...env.additionalKeys, ...state._volatile?.keys }
+        if (!state._vm.authorizedKeys) state._vm.authorizedKeys = {}
+        // Order is so that KEY_ADD doesn't overwrite existing keys
+        state._vm.authorizedKeys = { ...keysToMap(v), ...state._vm.authorizedKeys }
+        // TODO: Move to different function, as this is implemented in OP_CONTRACT as well
+        for (const key of v) {
+          if (key.meta?.private) {
+            if (key.id && key.meta.private.keyId in keys && key.meta.private.content) {
+              if (!state._volatile) state._volatile = { keys: {} }
+              try {
+                state._volatile.keys[key.id] = decrypt(keys[key.meta.private.keyId], key.meta.private.content)
+              } catch (e) {
+                console.error('Decryption error', e)
```

Comment:
This will display correctly in the browser console, but I think it won't display correctly in the user-accessible logs in the UI because the `e` object won't be properly translated to a string. So sometimes it's best to use `console.error` in a more explicit manner, like so:

```js
console.error(`OP_KEY_ADD decryption error '${e.message}':`, e)
```

Note also that this tells us where the error came from because it's a unique error message that doesn't appear anywhere else in the code.

So similar issues with the other `console.error`s elsewhere

Reply:
I would write it like so then:

```js
console.error(`OP_KEY_ADD decryption error '${e.message || e}':`, e)
```

As none of our code throws things that don't descend from `Error`, this shouldn't be an issue really.

Reply:
> As none of **our** code

It's really difficult to say with JS (well, C++ is strongly typed but you can also `throw` anything under the sun). I can however assume it'll always be an `Error`.

Reply:
Reminder!

Also same (new) question about throwing.